### PR TITLE
Add compute/storage protocol message size metrics

### LINKED ISSUE #16415: compute: Add protocol metrics
### Feature request

We currently don't collect any statistics of the environmentd-clusterd communication. These statics would help to identify network related problems and help us measure any progress we make in optimizing message sizes. In essence, it should provide a break down of the envd-computed bandwidth used. 

Part of #14663

## What metrics?

Since we create a network connection to each process in a replica, it makes most sense to collect these statistics per connection.

 * For each command type, a size histogram.
 * For each response type, a size histogram.

From these metrics we can then calculate total number of messages, message rate, total number of commands sent and responses received. 

## Open questions
 * What buckets for command/response sizes?
 * Check if our gRPC library [tonic](https://github.com/hyperium/tonic) exposes some of these stats already.

## Implementation

The least overhead approach would be to instrument out gRPC library tonic to report the size of the encoded/decoded message and store it. 

tonic allows to plug in custom Decoders (https://github.com/hyperium/tonic/pull/999). While writing a custom prost based protofbuf encoder is not very hard, the API makes it impossible to access (or pass in) the decoder instance from the outside. Thus we'd have to rely on some global variable or other hacks to access the statistics. Even if we write a custom encoder, we can't avoid some duplicate work, because the `encode` function of the prost message will always call `encoded_len` internally and there is no clean way to access it (unless we rely on the hidden API `encode_raw`).

### 1) Two `encoded_len` calls
However I think calling an extra `encoded_len` is acceptable. It will traverse the Protobuf twice, but given that we create two copies anyway (for example `ComputeCommand --> ProtoComputeCommand --> protobuf encoded buffer) an extra traversal seems not too expensive (also it will have perfect locality).

If we accept that we have to call `encode_len` twice, the implementation becomes easier. We can extend our `GrpcClient` to obtain statistics about messages and responses. 

### 2) Use [http_tower::trace](https://docs.rs/tower-http/latest/tower_http/trace/)

This crate intercepts the HTTP protocol. Because we are using a gRPC message stream, we only initiate one single long lived request. It is possible to inspect the *sent* body contents using the [OnBodyChunk](https://docs.rs/tower-http/latest/tower_http/trace/struct.TraceLayer.html#method.on_body_chunk) method. It's possible to count the total bytes sent using this method.

### 3) Don't use generated code for API calls

Since we have only one RPC in the service definition `command_response_stream` the tonic generated code is actually not very complicated. The wrapper around `command_response_stream` is 25 lines and we could inline those. That would allows us to inject a custom Codec object.

### LINKED ISSUE #16951: compute: Add replica connection metrics
### Feature request

We currently have to rely on the coarse kubernetes metrics or logs to diagnose any connection problems. This makes it hard to get a good overview of how long connection setup and replica initialization takes.

In particular we think to expose the following metrics:

* How long did it take to connect to replica?
* How often did it disconnect?
* How long does rehydration take?

### PR DESCRIPTION
**Pull request by lluki**
Add message size metric for messages.

Fixes #16415

### Motivation

  * This PR fixes a recognized bug 

### Tips for reviewer

I tried to make meaningful commits:

1st commit adds a mechanism to observe message sizes when they are sent via gRPC. Instead of using the tonic generated RPC stubs, we now interact directly with the tonic API. This allows us to provide a custom codec. Then we add a custom codec that collects metrics.

2nd commit uses that mechanism to expose metrics in our `MetricsRegistry`. I had to plumb it all the way down through the compute and storage controller stack.

3rd commit introduces the ability to provide custom buckets in `mz_ore::stats` and uses it to have meaningful size buckets for messages.

I did not yet refactor also the server side to not use the gRPC stubs, but I plan to refactor it as well to have client and server use the same mechanism.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - None

### CODE REVIEW

File: src/service/src/grpc.rs
Comment (guswynn) [inline src/service/src/grpc.rs:200]: this is pretty neat! much cleaner than before!

File: src/service/src/codec.rs
Comment (guswynn) [inline src/service/src/codec.rs:138]: pretty nice codec wrapper you made! pretty clean!

**Review by guswynn — COMMENTED** [review]
@lluki All the grpc changes seem good to me!

**Review by jkosh44 — APPROVED** [review]
surfaces owned files LGTM

**Comment by benesch** [comment]
I'm worried about the complexity of the interaction between the codecs and tonic. We've been fighting with gRPC quite a bit. I think it might be time to declare bankruptcy. This code would be far simpler, I think, if we just opened up a TCP socket and started firing length-delimited encoded protobufs across it. tokio-util has some helpers that make this easy.

cc @antiguru as someone familiar with gRPC from a past life, WDYT? Doesn't need to block this PR, necessarily, but I am worried about investing more time in our gRPC integration on the server side.

File: src/compute-client/src/metrics.rs
Comment (teskje) [inline src/compute-client/src/metrics.rs:101]: What's the purpose of the unused argument?

File: src/compute-client/src/metrics.rs
Comment (teskje) [inline src/compute-client/src/metrics.rs:112]: ISTM like everything would work without this type and we could use `ReplicaMetrics` directly to collect the metrics from the gRPC connection. Is there still a reason why we would want a separate type?

File: src/compute-client/src/metrics.rs
Comment (teskje) [inline src/compute-client/src/metrics.rs:119]: This warning is a bit confusing, since the range (`MIN..=MAX`) of `f64` is way larger than that of `u64`. Maybe we could clarify it to say that the integer value is not representable by `f64`?

Also, I think it would be a good idea to include the offending value in the log message.

File: src/controller/src/lib.rs
Comment (teskje) [inline src/controller/src/lib.rs:340]: It seems fitting to add this as a new field to `ControllerConfig` instead.

File: src/storage-client/src/metrics.rs
Comment (teskje) [inline src/storage-client/src/metrics.rs:25]: Copy-paste-o

File: src/storage-client/src/metrics.rs
Comment (teskje) [inline src/storage-client/src/metrics.rs]: With the cluster unification, the concept of a storage host has been removed (you'll see when you rebase). I think we can instead use the instance/replica concept here, like for the compute metrics.

**Comment by lluki** [comment]
> I'm worried about the complexity of the interaction between the codecs and tonic. We've been fighting with gRPC quite a bit. I think it might be time to declare bankruptcy. This code would be far simpler, I think, if we just opened up a TCP socket and started firing length-delimited encoded protobufs across it. tokio-util has some helpers that make this easy.

One benefit of gRPC is that we can use it to multiplex multiple streams. If we decide that at some point `clusterd` just offers one command endpoint while hosting multiple `storaged` or `computed` components, they could just get their own URL. And we can use standard tools to interact with it (ie. visualize communication in wireshark) - afaik noone actually used that though.

Instead of dropping the whole stack, we could also just drop tonic, but keep `tower_http` and lower layers to do `http/2` communication. That would give us for example multiplexing.

File: src/service/src/codec.rs
Comment (lluki) [reply src/service/src/codec.rs:60]: We'd need to turn it into a gRPC status which is described as "A gRPC status describing the result of an RPC call.". I'm not sure that's a perfect match. This is also the same what tonic does internally (so it's the behavior we currently have).

File: src/service/src/codec.rs
Comment (lluki) [reply src/service/src/codec.rs:104]: Thanks, don't know what I was thinking

File: src/compute-client/src/metrics.rs
Comment (lluki) [reply src/compute-client/src/metrics.rs:101]: If we have a multi process replica, we could differentiate on the process. But we currently agreed on collecting the metrics per-replica (see [slack discussion](https://materializeinc.slack.com/archives/CM7ATT65S/p1673884428820379)).

File: src/compute-client/src/metrics.rs
Comment (lluki) [reply src/compute-client/src/metrics.rs:112]: Not as of now, but if we want to add the more detailed per-process metrics, it would be helpful. I remove it.

File: src/service/src/codec.rs
Comment (teskje) [reply src/service/src/codec.rs:60]: I guess the "Internal" status could be used as a catch-all, but it's probably fine to panic when we are about to OOM anyway.

**Comment by tokenrove** [comment]
> We've been fighting with gRPC quite a bit. I think it might be time to declare bankruptcy. This code would be far simpler, I think, if we just opened up a TCP socket and started firing length-delimited encoded protobufs across it. tokio-util has some helpers that make this easy.

I'm a big fan of this, I just want to add the caveat to take into account whether things like head-of-line blocking can happen and need to be handled, since normally HTTP/2 under the hood is going to be facilitating some of this for gRPC, at the cost of a huge amount of complexity underneath.  Something message-oriented like QUIC or even the venerable SCTP might be worth considering instead of TCP, too, since we have clear message boundaries.

**Comment by benesch** [comment]
> > We've been fighting with gRPC quite a bit. I think it might be time to declare bankruptcy. This code would be far simpler, I think, if we just opened up a TCP socket and started firing length-delimited encoded protobufs across it. tokio-util has some helpers that make this easy.
> 
> I'm a big fan of this, I just want to add the caveat to take into account whether things like head-of-line blocking can happen and need to be handled, since normally HTTP/2 under the hood is going to be facilitating some of this for gRPC, at the cost of a huge amount of complexity underneath. Something message-oriented like QUIC or even the venerable SCTP might be worth considering instead of TCP, too, since we have clear message boundaries.

Is head-of-line blocking an issue given the way the protocol is designed, though? The responses are asynchronous; a given request doesn't block for a response; it's just you expect to maybe see a response addressing your request at some point in the future.

**Comment by lluki** [comment]
> Is head-of-line blocking an issue given the way the protocol is designed, though? The responses are asynchronous; a given request doesn't block for a response; it's just you expect to maybe see a response addressing your request at some point in the future.

My biggest concern is, for example if we send a plan with many constants to compute, we don't want to slow down messages meant for storage. But we'd be using http/2 which interleaves multiple streams and solves that problem. 

I had to read up what HOL blocking actually is in http/2 ([link](https://calendar.perfplanet.com/2020/head-of-line-blocking-in-quic-and-http-3-the-details/)). It's the TCP retransmission that happen on packet loss which will block all the streams instead of just the one affected by packet loss. Given that this is a in-datacenter communication, packet loss rate should be small and our commands also tend to be small (although we will now more once this is merged), I think this is not a big issue in our case.

File: src/ore/src/metrics.rs
Comment (teskje) [inline src/ore/src/metrics.rs:87]:
```suggestion
            buckets: None,
```

**Review by teskje — APPROVED** [review]
LGTM!

File: src/ore/src/metrics.rs
Comment (lluki) [reply src/ore/src/metrics.rs]: ty

**Comment by teskje** [comment]
Re the gRPC discussion: ISTM that the only point of contention of replacing gRPC with something much simpler (like TCP or QUIC) is that we will run into issues trying to multiplex streams. But I'm not sure why we would even want to do that. As long as the timely runtimes are still separate, I don't think there is any harm in having a compute and a storage server running on different ports, like we do now. Would multiplexing give us anything in this scenario, apart from maybe slightly simplified operations due to having to manage one less port?

I'm less clear on what we want once we merge both timely runtimes. My intuition is that we would want compute and storage commands to arrive in the same interleaving order as they are sent by the controller(s), to ensure determinism, which might or might not be important for the different timely processes to communicate successfully*, and might or might not be important for us to maintain sanity. If this is indeed the case, multiplexing would not be useful to us.

-------

\* [This comment](https://github.com/MaterializeInc/materialize/issues/16773#issuecomment-1368012916) by @aljoscha points out that:

> In order for communication to work correctly between workers, timely dataflows/operators must be constructed in the same (deterministic) order on all workers. The handwave-y explanation is that timely internally uses an incrementing ID for identifying operators.

Based on that, it would be bad if in a multi-process replica, one process would create a storage source before a compute dataflow, and another process would create both in reverse order.

**Comment by lluki** [comment]
> I'm less clear on what we want once we merge both timely runtimes. My intuition is that we would want compute and storage commands to arrive in the same interleaving order as they are sent by the controller(s), to ensure determinism, 

Ah right, there is no point in trying to send compute/storage commands on separate channels, they must be ordered. Unless we make a distinction in commands that change the timely state and those that don't, but I don't think there's much to be gained... At which point yes, we could just open a TCP socket and send length delimited protobuf messages (which is what the current gRPC async stream is doing after setup anyway).

**Comment by benesch** [comment]
> > I'm less clear on what we want once we merge both timely runtimes. My intuition is that we would want compute and storage commands to arrive in the same interleaving order as they are sent by the controller(s), to ensure determinism,
> 
> Ah right, there is no point in trying to send compute/storage commands on separate channels, they must be ordered. Unless we make a distinction in commands that change the timely state and those that don't, but I don't think there's much to be gained... At which point yes, we could just open a TCP socket and send length delimited protobuf messages (which is what the current gRPC async stream is doing after setup anyway).

Exactly! I filed #17375 to track the work item.

**Pull request merged**
